scanpy/preprocessing/simple.py

```diff
@@ -740,52 +743,78 @@ def regress_out(adata, keys, n_jobs=None, copy=False):
+    chunk_list = np.array_split(adata.X, n_chunks, axis=1)
+    if variable_is_categorical:
+        regressors_chunk = np.array_split(regressors, n_chunks, axis=1)
+    for idx, data_chunk in enumerate(chunk_list):
+        # each task is a tuple of a data_chunk eg. (adata.X[:,0:100]) and
+        # the regressors. This data will be passed to each of the jobs.
+        if variable_is_categorical:
+            regres = regressors_chunk[idx]
+        else:
+            regres = regressors
+        tasks.append(tuple((data_chunk, regres, variable_is_categorical)))
+
+    if n_jobs > 1 and n_chunks > 1:
+        import multiprocessing
+        pool = multiprocessing.Pool(n_jobs)
+        res = pool.map_async(_regress_out_chunk, tasks).get(9999999)
+        pool.close()
+
+    else:
+        res = list(map(_regress_out_chunk, tasks))
+
+    # res is a list of vectors (each corresponding to a regressed gene column).
+    # The transpose is needed to get the matrix in the shape needed
+    adata.X = np.vstack(res).T.astype(adata.X.dtype)
+    logg.info('    finished', t=True)
+    return adata if copy else None
+
+
+def _regress_out_chunk(data):
+    # data is a tuple containing the selected columns from adata.X
+    # and the regressors dataFrame
+    data_chunk = data[0]
+    regressors = data[1]
+    variable_is_categorical = data[2]
+
+    responses_chunk_list = []
     import statsmodels.api as sm
     from statsmodels.tools.sm_exceptions import PerfectSeparationError
 
-    def _regress_out(col_index, responses, regressors):
+    for col_index in range(data_chunk.shape[1]):
+        if variable_is_categorical:
+            regres = np.c_[np.ones(regressors.shape[0]), regressors[:, col_index]]
+        else:
+            regres = regressors
         try:
-            if regressors.shape[1] - 1 == responses.shape[1]:
-                regressors_view = np.c_[regressors[:, 0], regressors[:, col_index + 1]]
-            else:
-                regressors_view = regressors
-            result = sm.GLM(responses[:, col_index],
-                            regressors_view, family=sm.families.Gaussian()).fit()
+            result = sm.GLM(data_chunk[:, col_index], regres, family=sm.families.Gaussian()).fit()
             new_column = result.resid_response
         except PerfectSeparationError:  # this emulates R's behavior
             logg.warn('Encountered PerfectSeparationError, setting to 0 as in R.')
-            new_column = np.zeros(responses.shape[0])
-        return new_column
-
-    def _regress_out_chunk(chunk, responses, regressors):
-        chunk_array = np.zeros((responses.shape[0], chunk.size),
-                               dtype=responses.dtype)
-        for i, col_index in enumerate(chunk):
-            chunk_array[:, i] = _regress_out(col_index, responses, regressors)
-        return chunk_array
-
-    for chunk in chunks:
-        # why did this break after migrating to dataframes?
-        # result_lst = Parallel(n_jobs=n_jobs)(
```

Comment:
Hi @fidelram! Thank you very much for the PR! Absolutely awesome! Also with the benchmarks... Just one question: Does your solution not have the memory issues that I encountered here? I'm just a bit worried that things will crash when you pass, say a list of 20 genes as regressors. If you're confident that this is fine, I'm happy to merge this immediately.

Reply:
In the current implementation, each processor receives a) a copy of the adata.X matrix containing **only** the columns to regress and b) the columns used as independent variables. 

a) is efficient because only the portion needed is passed
b) is copied multiple times as jobs. If b is big this could take some memory but not much. The size of b is the number of cells x num. of independent variables. Even if the num of independent variables is 20, this should be too big. Let me run some tests to check this out.

Reply:
I forgot to mention that if a categorical variable is used, in which case a matrix of size equal to adata.X is generated, 'b' (from the above comment) is also efficient because only the part of the matrix needed is passed to each processor.